### PR TITLE
Refactor InboundHandler into smaller Methods

### PR DESCRIPTION
**Pull request by original-brownbear**
I tried to go through this code with someone and it became evident that the massive method for request handling is very hard to grok. I split it up into its logical constituents here and fixed some redundant passing around of `message` and `header`.
This at least makes it so we only have one layer of ref count handling per method and it is a little easier to follow where forking happens. Also, this removes a weirdly long multi-second compile task from C2 for whatever that's worth :)

Also:
* fixed `assertRemoteVersion` to be a proper assertion method
* extracted `verifyRequestReadFully` since it's not essential to the logic just like we have `verifyResponseReadFully`

### CODE REVIEW

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-1

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java [hunk 1]
```diff
@@ -106,70 +106,98 @@ private void messageReceived(TcpChannel channel, InboundMessage message, long st
...
+            header.getRequestId(),
+            messageListener
+        );
+        final TransportResponseHandler<?> responseHandler;
+        if (theHandler == null && header.isError()) {
+            responseHandler = handshaker.removeHandlerForHandshake(header.getRequestId());
+        } else {
+            responseHandler = theHandler;
+        }
+        return responseHandler;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundHandler.java:169]: nit: no need for local `responseHandler` variable, we can just return directly.

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java [hunk 2]
```diff
@@ -193,139 +221,152 @@ private void verifyResponseReadFully(Header header, TransportResponseHandler<?>
...
-                                }
-                            }
-                        }
-                    } finally {
-                        request.decRef();
-                    }
+    private static <T extends TransportRequest> void doHandleRequest(
+        RequestHandlerRegistry<T> reg,
+        T request,
+        TransportChannel transportChannel
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundHandler.java:291]: nit: rename to `channel` like with `handleRequestForking`, then the method signature fits on one line.

File: server/src/main/java/org/elasticsearch/transport/InboundHandler.java [hunk 3]
```diff
@@ -106,70 +106,98 @@ private void messageReceived(TcpChannel channel, InboundMessage message, long st
...
-                        took,
-                        logThreshold
-                    );
-                }
+                logSlowMessage(message, took, logThreshold, responseHandler);
+            }
+        }
+    }
+
+    private void resolveResponseHandler(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/transport/InboundHandler.java:135]: nit: `resolve` suggests to me we're doing some kind of lookup to find the response handler, but in fact we already have the handler and this method is running it. Which admittedly could be called a "resolution". But still, I'd rather name this something like `completeResponseHandler` or `executeResponseHandler`.

**Review by DaveCTurner — APPROVED** [review]
LGTM, I left a few superficial suggestions. It's easier to review this kind of thing as a sequence of commits tho, I ended up having to repeat the refactoring myself to be sure it all made sense.

**Comment by original-brownbear** [comment]
Thanks David!

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/part-1

**Pull request merged**